### PR TITLE
Anthropic (Claude) provider

### PR DESCRIPTION
**Pull request by brichet**
This PR adds the Anthropic provider as a chat model.
~~It does not include completion.~~

### CODE REVIEW

**Comment by brichet** [comment]
I added a first trial on code completion, the idea is to give a context to the AI, and to send the current prompt as a message to complete.
The 2 issues I have so far are:
- it is really slow on my side
- anthropic raises if there is a space at the end of the prompt. Currently this PR is removing it, but it can lead to wrong suggestions

**Comment by jtpio** [comment]
Thanks @brichet for looking into this :+1: 

I just tried this PR locally and there seems to be many requests made to the anthropic API, so maybe there is indeed something not expected happening:

[jupyterlite-anthropic.webm](https://github.com/user-attachments/assets/c079b941-1557-401c-9249-4b156a936151)

**Comment by jtpio** [comment]
Interestingly when testing with Jupyter AI, completions don't seem to work and trigger some warnings in the dev tools console:

[jupyter-ai-lab-434-anthropic.webm](https://github.com/user-attachments/assets/2e602d4c-325d-4346-803b-b09e556fbda3)


The Jupyter Server logs also show some long response time for the inline completer:

image

**Comment by brichet** [comment]
Thanks for testing it @jtpio

Using Anthropic completion with jupyter-ai do not trigger error on my side, but the suggestions are not really reliable (most of the time):

Peek 2024-12-20 13-59


Indeed, you're right about the extra requests, I'll take a closer look.

**Comment by jtpio** [comment]
> Using Anthropic completion with jupyter-ai do not trigger error on my side, but the suggestions are not really relevant (most of the time):

Indeed according to your screenshot there seems to be an issue with how Jupyter AI handles the inline completer suggestions. So likely an issue in Jupyter AI (we could open one if non exists already)

**Comment by brichet** [comment]
Looks like the default model is `claude-2.1`, which seems to be slower and less accurate than `claude-3-5-sonnet-20241022`. Setting it manually improve the performances of the code completion.

Maybe we should revive https://github.com/jupyterlite/ai/pull/14 to get all the settings for each providers.

**Comment by jtpio** [comment]
> Maybe we should revive #14 to get all the settings for each providers.

Yeah the JupyterLab related change should now be available in `4.4.0a2`.

**Comment by jtpio** [comment]
> Looks like the default model is `claude-2.1`, which seems to be slower and less accurate than `claude-3-5-sonnet-20241022`. Setting it manually improve the performances of the code completion.

Indeed, I just tried it locally and it seems to be faster with Claude 3.5 Sonnet.

There seems to be some issues with the quality of completion being returned sometimes, but it's something that can likely be fixed by improving the prompt or the use of the langchain primitives:

image

**Comment by brichet** [comment]
Unfortunately https://github.com/jupyterlite/ai/pull/14 does not work out of the box for Anthropic provider.

**Comment by jtpio** [comment]
> Unfortunately #14 does not work out of the box for Anthropic provider.

Looks like the settings generator script needs to be updated for each new provided being added?

**Comment by brichet** [comment]
> Looks like the settings generator script needs to be updated for each new provided being added?

Yes, but it would be great if we could have a generic way to retrieve the available settings.

If `@langchainjs` is using a naming convention for the modules and the "input" interfaces we can make it more generic, but it seems to not be the case (eg. `ChatMistralAIInput` and `AnthropicInput`).

The other problem is about the use of [ts-json-schema-generator](https://github.com/vega/ts-json-schema-generator), which raises an error when trying to build the settings for `Anthropic` without excluding `clientOptions` (this king of error seems to not be handled correctly): 

```
/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeReferenceNodeParser.js:59
        return this.childNodeParser.createType(typeSymbol.declarations.filter((n) => !invalidTypes[n.kind])[0], this.createSubContext(node, context));
                                                                       ^

TypeError: Cannot read properties of undefined (reading 'filter')
    at TypeReferenceNodeParser.createType (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeReferenceNodeParser.js:59:72)
    at ChainNodeParser.createType (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:30:45)
    at TypeReferenceNodeParser.createSubContext (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeReferenceNodeParser.js:65:62)
    at TypeReferenceNodeParser.createType (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeReferenceNodeParser.js:59:118)
    at ChainNodeParser.createType (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:30:45)
    at TypeAliasNodeParser.createType (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/NodeParser/TypeAliasNodeParser.js:36:43)
    at AnnotatedNodeParser.createType (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/NodeParser/AnnotatedNodeParser.js:29:47)
    at ExposeNodeParser.createType (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/ExposeNodeParser.js:24:45)
    at CircularReferenceNodeParser.createType (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/CircularReferenceNodeParser.js:22:43)
    at ChainNodeParser.createType (/home/brichet/projects/jupyterlite-ai/node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:30:45)
```

In conclusion, for now I think we'll have to update the generator, but hopefully it should be painless next time :smile:

File: scripts/settings-generator.js [hunk 1]
```diff
@@ -20,23 +20,57 @@ const schemaBase = tsj
...
   };
 
   const outputPath = path.join(outputDir, `${name}.json`);
 
-  const schema = tsj.createGenerator(config).createSchema(config.type);
+  const generator = tsj.createGenerator(config);
+  let schema;
+
+  // Workaround to exclude some properties from a type or interface.
+  if (desc.excludedProps) {
```
Comment (jtpio) [inline scripts/settings-generator.js:61]: It's probably fine to have exclusion clauses like this for now as we figure things out with several providers 👍 

Also found that some props had to be excluded over in https://github.com/jupyterlite/ai/pull/27.

File: src/llm-models/anthropic-completer.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    request: CompletionHandler.IRequest,
+    context: IInlineCompletionContext
+  ) {
+    const { text, offset: cursorOffset } = request;
+    const prompt = text.slice(0, cursorOffset);
+
+    // Anthropic does not allow whitespace at the end of the AIMessage
+    const trimmedPrompt = prompt.trim();
+
+    const messages = [
```
[the remark this answers is not in the document]
Comment (jtpio) [reply src/llm-models/anthropic-completer.ts:30]: Some other open questions: should the list of messages be shared across providers? Or should it be reset when a user switches to another provider?

Maybe a reset would be fine to keep things tidy. Also users probably wouldn't change providers often in practice?

**Review by jtpio — APPROVED** [review]
Thanks!

Testing locally and it seems to be working fine 👍 

By default the inline completion does seem to be quite slow:

https://github.com/user-attachments/assets/ec6fbfc5-b5af-4ab7-a558-7a41844276e7


But switching to `claude-3-5-sonnet-latest` as the model does seem to be making things faster:

https://github.com/user-attachments/assets/acdb9acd-7d5a-4575-9159-0c94b693bb0d


Left a couple of inline comments, that we can also track separately.

File: src/llm-models/anthropic-completer.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jtpio) [reply src/llm-models/anthropic-completer.ts:30]: Opened https://github.com/jupyterlite/ai/issues/29

**Comment by brichet** [comment]
Thanks for the review @jtpio
